### PR TITLE
SALTO-6867 Fetch bot builder

### PR DESCRIPTION
**Pull request by adfineberg**
This is still missing a filter to order the paths, some references and error handling for the graphql response

---

_Additional context for reviewer_

---
_Release Notes_: 
_Replace me with a short sentence that describes the effect of this change on Salto users_

---
_User Notifications_: 
_Replace me with a short sentence that describes changes that will appear in NaCls and are not caused by user actions (e.g. a new annotation, field values that are converted to references, etc). Hidden changes should not be listed._

### CODE REVIEW

**Review by edenhassid — COMMENTED** [review]
looks good! 🔥 added some small comments

File: packages/zendesk-adapter/src/definitions/requests/clients.ts [hunk 1]
```diff
@@ -80,5 +80,16 @@ export const createClientDefinitions = (
         },
       },
     },
+    graphql: {
+      httpClient: clients.graphql,
+      endpoints: {
+        default: {
+          post: {
+            // This is for the graphql endpoint, which is readonly for fetch. It allows it to be used in fetch
```
Comment (netama) [inline packages/zendesk-adapter/src/definitions/requests/clients.ts:88]: then maybe set it specifically for the graphql path? or are there multiple?
(is there a reason to do this in a separate PR from the fetch definitions?)

File: packages/zendesk-adapter/src/adapter.ts [hunk 2]
```diff
@@ -555,8 +555,11 @@ export default class ZendeskAdapter implements AdapterOperations {
       }
       return clientsBySubdomain[subdomain]
     }
-
-    this.fetchQuery = elementUtils.query.createElementQuery(this.userConfig[FETCH_CONFIG], fetchCriteria)
+    const fetchConfig = this.userConfig[FETCH_CONFIG]
+    if (this.userConfig[FETCH_CONFIG].fetchBotBuilder === false) {
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/adapter.ts:559]:
```suggestion
    if (fetchConfig.fetchBotBuilder === false) {
```

File: packages/zendesk-adapter/src/adapter.ts [hunk 3]
```diff
@@ -555,8 +555,11 @@ export default class ZendeskAdapter implements AdapterOperations {
...
+      fetchConfig.exclude.push({ type: 'bot_builder_flow' })
+    }
+    this.fetchQuery = elementUtils.query.createElementQuery(fetchConfig, fetchCriteria)
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/adapter.ts:562]: why do you do this?

File: packages/zendesk-adapter/src/definitions/fetch/fetch.ts [hunk 4]
```diff
@@ -1842,6 +1849,86 @@ const createCustomizations = (): Record<
...
+      },
+    },
+  },
+  bot_builder_node: {
+    element: {
+      topLevel: {
+        isTopLevel: true,
+        elemID: {
+          parts: [
+            { fieldName: 'externalId', condition: value => value?.externalId && value?.externalId !== null },
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/definitions/fetch/fetch.ts:1917]: why are we using the external id?

File: packages/zendesk-adapter/src/definitions/fetch/fetch.ts [hunk 5]
```diff
@@ -1842,6 +1849,86 @@ const createCustomizations = (): Record<
...
+            { fieldName: 'id', condition: value => !(value?.externalId && value?.externalId !== null) },
+          ],
+          extendsParent: true,
+        },
+        path: {
+          pathParts: [{ parts: [{ fieldName: 'name' }], extendsParent: true }],
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/definitions/fetch/fetch.ts:1923]: name is correct here?

File: packages/zendesk-adapter/src/definitions/fetch/fetch.ts [hunk 6]
```diff
@@ -1842,6 +1849,86 @@ const createCustomizations = (): Record<
...
+      },
+    },
+  },
+  bot_builder_answer: {
+    element: {
+      topLevel: {
+        isTopLevel: true,
+        elemID: { parts: DEFAULT_ID_PARTS, extendsParent: true },
+        path: {
+          pathParts: [{ parts: [{ fieldName: 'name' }], extendsParent: true }],
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/definitions/fetch/fetch.ts:1895]: why not use DEFAULT_ID_PARTS?

File: packages/zendesk-adapter/src/definitions/fetch/fetch.ts [hunk 7]
```diff
@@ -1842,6 +1849,86 @@ const createCustomizations = (): Record<
...
+            { fieldName: 'id', condition: value => !(value?.externalId && value?.externalId !== null) },
+          ],
+          extendsParent: true,
+        },
+        path: {
+          pathParts: [{ parts: [{ fieldName: 'name' }], extendsParent: true }],
+        },
+      },
+      fieldCustomizations: {
+        id: { fieldType: 'string', hide: true },
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/definitions/fetch/fetch.ts:1927]: I don't think we need to hide it, WDYT?

File: packages/zendesk-adapter/src/definitions/fetch/transforms/graphql_adjuster.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+export const transform: (innerRoot: string) => definitions.AdjustFunctionMulti =
+  (innerRoot: string) =>
+  async ({ value }) => {
+    if (!lowerdashValues.isPlainObject(value)) {
+      throw new Error('unexpected value for graphql item, not transforming')
+    }
+    const graphqlValues = get(value, `data.${innerRoot}`)
+    return graphqlValues === undefined
+      ? { value }
+      : graphqlValues.map((graphqlValue: Record<string, unknown>) => ({
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/definitions/fetch/transforms/graphql_adjuster.ts:22]: not sure i understand what is happening here, you get a list of flows and you return a list of values? this is why you use `AdjustFunctionMulti`?

File: packages/zendesk-adapter/test/adapter.test.ts [hunk 9]
```diff
@@ -777,6 +782,21 @@ describe('adapter', () => {
...
+          'zendesk.bot_builder_flow__assignedChannelIntegrations',
+          'zendesk.bot_builder_flow__fallback',
+          'zendesk.bot_builder_flow__freeTextQuery',
+          'zendesk.bot_builder_flow__greeting',
+          'zendesk.bot_builder_flow__greeting__suggestedAnswers',
+          'zendesk.bot_builder_flow__helpCenterAutoReplyFeedback',
+          'zendesk.bot_builder_flow__publishedChannelIntegrations',
+          'zendesk.bot_builder_node',
+          'zendesk.bot_builder_node.instance.myBrand_test_bot_us__Testing_uumuu__01J4SRF1HAS4TQFQ99PKFRCQXB@uuuuumuu',
+          'zendesk.bot_builder_node__data',
```
Comment (edenhassid) [inline packages/zendesk-adapter/test/adapter.test.ts:799]: remove the 'uumuu' in the middle, there is a flag in the definitions that does that I think

**Review by edenhassid — COMMENTED** [review]
looks great!! 🎉 💪  added some small comments :)

File: packages/zendesk-adapter/test/adapter.test.ts (hunk 9 above)
Comment (adfineberg) [reply packages/zendesk-adapter/test/adapter.test.ts:799]: Sadly, because of the deprecated filter `referencedInstanceNamesFilterCreatorDeprecated` nested files that have an elemID change in the filter have this in the old infra :(

File: packages/zendesk-adapter/src/adapter.ts (hunk 3 above)
Comment (adfineberg) [reply packages/zendesk-adapter/src/adapter.ts:562]: discussed f2f

File: packages/zendesk-adapter/src/definitions/fetch/fetch.ts (hunk 5 above)
Comment (adfineberg) [reply packages/zendesk-adapter/src/definitions/fetch/fetch.ts:1920]: fixed

**Review by edenhassid — APPROVED** [review]
LGTM! 🔥
